scripts/controlnet.py

```diff
@@ -261,420 +165,20 @@ def get_default_ui_unit(self, is_ui=True):
-            return input_mode, UiControlNetUnit(*args)
-
-        for input_tab in (
-            (upload_tab, batch_hijack.InputMode.SIMPLE),
-            (batch_tab, batch_hijack.InputMode.BATCH)
-        ):
-            input_tab[0].select(fn=ui_controlnet_unit_for_input_mode, inputs=[gr.State(input_tab[1])] + list(unit_args), outputs=[input_mode, unit])
-
-        def determine_batch_dir(batch_dir, fallback_dir, fallback_fallback_dir):
-            if batch_dir:
-                return batch_dir
-            elif fallback_dir:
-                return fallback_dir
-            else:
-                return fallback_fallback_dir
-
-        # keep batch_dir in sync with global batch input textboxes
-        global img2img_batch_input_dir, img2img_batch_input_dir_callbacks
-        def subscribe_for_batch_dir():
-            global global_batch_input_dir, img2img_batch_input_dir
-            batch_dirs = [batch_image_dir, global_batch_input_dir, img2img_batch_input_dir]
-            for batch_dir_comp in batch_dirs:
-                subscriber = getattr(batch_dir_comp, 'blur', None)
-                if subscriber is None: continue
-                subscriber(
-                    fn=determine_batch_dir,
-                    inputs=batch_dirs,
-                    outputs=[batch_image_dir_state],
-                    queue=False,
-                )
-
-        if img2img_batch_input_dir is None:
-            # we are too soon, subscribe later when available
-            img2img_batch_input_dir_callbacks.append(subscribe_for_batch_dir)
-        else:
-            subscribe_for_batch_dir()
-
-        # keep output_dir in sync with global batch output textbox
-        global img2img_batch_output_dir, img2img_batch_output_dir_callbacks
-        def subscribe_for_output_dir():
-            global img2img_batch_output_dir
-            img2img_batch_output_dir.blur(
-                fn=lambda a: a,
-                inputs=[img2img_batch_output_dir],
-                outputs=[output_dir_state],
-                queue=False,
-            )
-
-        if img2img_batch_input_dir is None:
-            # we are too soon, subscribe later when available
-            img2img_batch_output_dir_callbacks.append(subscribe_for_output_dir)
-        else:
-            subscribe_for_output_dir()
-
-        if is_img2img:
-            img2img_submit_button.click(fn=UiControlNetUnit, inputs=list(unit_args), outputs=unit, queue=False)
-        else:
-            txt2img_submit_button.click(fn=UiControlNetUnit, inputs=list(unit_args), outputs=unit, queue=False)
-
-        return unit
+    def uigroup(self, tabname: str, is_img2img: bool, elem_id_tabname: str):
+        group = ControlNetUnit(
+            gradio_compat,
+            self.infotext_fields,
+            self.get_default_ui_unit(),
+            self.preprocessor,
+            self.txt2img_w_slider,
+            self.txt2img_h_slider,
+            self.img2img_w_slider,
+            self.img2img_h_slider,
```

Comment:
We could move these sliders directly into the new `ControlNetUnit` class, no need to pass them here I believe. To register them, use the same function as for the batch input/output text boxes.